server-spi-private/src/main/java/org/keycloak/models/locking/GlobalLock.java

```diff
@@ -0,0 +1,35 @@
+/*
+ * Copyright 2022 Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.keycloak.models.locking;
+
+/**
+ * An object of this type represents a successfully acquired global lock provided by {@link GlobalLockProvider}
+ */
+public interface GlobalLock extends AutoCloseable {
+
+    /**
+     * Releases the lock represented by this
+     */
+    @Override
+    void close();
+
+    public static class Constants {
+        public static final String DATABASE = "database";
```

Comment:
What are the types of server-wide locks we have? I can think about a lock during boot (`keycloak-boot`). Any other?

The reference to `database` there reads a bit weird as it is more an implementation detail on how the lock is implemented, right?

Reply:
> What are the types of server-wide locks we have? I can think about a lock during boot (keycloak-boot). Any other?

For example, we need to lock specific area when initializing new map storage jpa, see: https://github.com/keycloak/keycloak/blob/main/model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/JpaMapStorageProviderFactory.java#L436

Another example is locking HotRod storage when the proto schema is registered with the Infinispan server. Having more granular locks allows us to leave part of Keycloak operational while some other part is initialized.

> The reference to database there reads a bit weird as it is more an implementation detail on how the lock is implemented, right?

The constant `database` in this case doesn't mean the locks are provided by the database. The constant serves for marking all blocks of the code that should not be visited by more threads/nodes at the same time.

So for example, if Liquibase update is running we will acquire lock with name `database`, this means that all other threads/nodes know someone else is doing Liquibase update so they wait until the lock is released.

Reply:
> For example, we need to lock specific area when initializing new map storage jpa, see: https://github.com/keycloak/keycloak/blob/main/model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/JpaMapStorageProviderFactory.java#L436
> 
> Another example is locking HotRod storage when the proto schema is registered with the Infinispan server. Having more granular locks allows us to leave part of Keycloak operational while some other part is initialized.
>

If both locks you mentioned are related to migration/schema operations, looks like we have a specific type of lock similarly to what you did to `keycloak-boot`. Perhaps `keycloak-init`or something. Or even different ones specific for `keycloak-migration` and `keycloak-init`.
 
> > The reference to database there reads a bit weird as it is more an implementation detail on how the lock is implemented, right?
> 
> The constant `database` in this case doesn't mean the locks are provided by the database. The constant serves for marking all blocks of the code that should not be visited by more threads/nodes at the same time.
> 

But `database` implies it is somehow related to a database. The whole purpose of having a "global" lock is do to what you mentioned so it is still not clear why we need `database`. Sorry, if I'm missing something.

> So for example, if Liquibase update is running we will acquire lock with name `database`, this means that all other threads/nodes know someone else is doing Liquibase update so they wait until the lock is released.

I think a more generic approach and also more aligned with the different usages of locks is to not use `database`. In this example, you are talking about Liquibase and we could just have a constant similarly to `keycloak-boot` but specific for schema init/migration, as mentioned before.

Reply:
> If both locks you mentioned are related to migration/schema operations, looks like we have a specific type of lock similarly to what you did to keycloak-boot. Perhaps keycloak-initor something. Or even different ones specific for keycloak-migration and keycloak-init.

The locks I mentioned are not one specific lock, it can be lock per storage area. You can lock only realms or only users. You don't need to lock the whole database when initializing, for example, authz services, other areas can continue working.

> But database implies it is somehow related to a database. The whole purpose of having a "global" lock is do to what you mentioned so it is still not clear why we need database. Sorry, if I'm missing something.

Maybe I am missing something as well, what is your suggestion here? 

> I think a more generic approach and also more aligned with the different usages of locks is to not use database. In this example, you are talking about Liquibase and we could just have a constant similarly to keycloak-boot but specific for schema init/migration, as mentioned before.

> If you are using database or hot rod locks, it does not matter. But that we are using a locking mechanism for boot, init, or migration.

This is exactly what is this PR about, we changed the locks to not provide only three specific locks as it was before. Now it is possible to create a lock with any String identifier, so you can create, for example, lock for only one user, or for only one storage area, or whatever you currently need to synchronize across your deployment. 

These constants here are not meant to somehow restrict what lock names can be used, it is there just to make sure we use the same String in all places we want to use this specific lock and also to easily find all places that are synchronized for this String.

Reply:
> > But database implies it is somehow related to a database. The whole purpose of having a "global" lock is do to what you mentioned so it is still not clear why we need database. Sorry, if I'm missing something.
> 
> Maybe I am missing something as well, what is your suggestion here?

I'm proposing removing those constants. Mainly `database` as it is not aligned with what you are saying. For instance, `database` is an implementation detail, and having it on an interface does not make much sense.

Have specific methods on the provider interface to return a specific type of lock such as a boot, init migration, user, etc, lock. For me, it makes sense to make explicit what types of locks you can use.

Something like:

```
LockProvider provider = // get provider
Lock lock1 = provider.getServerBootLock();
Lock lock2 = provider.getInitLock();
Lock lock3 = provider.getMigrationLock();
StorageLock storageLock = provider.getStorageLock();
Lock lock = storageLock.getUserLock(area);
```
No reference to impl details (like database) but making the possible types of locks more explicit in the provider contract.

Reply:
I believe the `DATABASE` lock can be moved fully into legacy-private, correct?
Then the only lock string which should remain is the `KEYCLOAK_BOOT` lock.

As for individual methods per lock type, these go against the purpose of this PR. Various storages need various locks (e.g. `DATABASE` and `MIGRATION` are only relevant for legacy store; there are also additional upgrading lock types anticipated in map storage), and they both cannot be listed in advance and cannot be limited by API. This is the main reason for having the lock ID as a free `String` parameter.

The same applies to future background tasks which will also need locks for their ability to operate, and we do not know the their full list as they will be created upon demand.

For both reasons, the API should not be fixed by method names.

Reply:
> I believe the `DATABASE` lock can be moved fully into legacy-private, correct? Then the only lock string which should remain is the `KEYCLOAK_BOOT` lock.

Sounds better so we make the new SPI more aligned with the expected design.

> For both reasons, the API should not be fixed by method names.

Ok.

Reply:
@hmlnarik It seems `DATABASE` constant can be moved to the legacy module. But we would need to change the code here: 
https://github.com/keycloak/keycloak/blob/main/model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/JpaMapStorageProviderFactory.java#L438

however, it is probably the correct thing to do in this PR anyway. (replace `DATABASE` with `modelType`)